java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowStreamReader.java

```diff
@@ -106,20 +108,42 @@ public boolean loadNextBatch() throws IOException {
       return false;
     }
 
-    if (result.getMessage().headerType() != MessageHeader.RecordBatch) {
-      throw new IOException("Expected RecordBatch but header was " + result.getMessage().headerType());
+    if (result.getMessage().headerType() == MessageHeader.RecordBatch) {
+      ArrowBuf bodyBuffer = result.getBodyBuffer();
+
+      // For zero-length batches, need an empty buffer to deserialize the batch
+      if (bodyBuffer == null) {
+        bodyBuffer = allocator.getEmpty();
+      }
+
+      ArrowRecordBatch batch = MessageSerializer.deserializeRecordBatch(result.getMessage(), bodyBuffer);
+      loadRecordBatch(batch);
+      checkDictionaries();
+      return true;
+    } else if (result.getMessage().headerType() == MessageHeader.DictionaryBatch) {
+      // if it's dictionary message, read dictionary message out and continue to read unless get a batch or eos.
+      ArrowDictionaryBatch dictionaryBatch = readDictionary(result);
+      loadDictionary(dictionaryBatch);
+      return loadNextBatch();
+    } else {
+      throw new IOException("Expected RecordBatch or DictionaryBatch but header was " +
+          result.getMessage().headerType());
     }
+  }
 
-    ArrowBuf bodyBuffer = result.getBodyBuffer();
-
-    // For zero-length batches, need an empty buffer to deserialize the batch
-    if (bodyBuffer == null) {
-      bodyBuffer = allocator.getEmpty();
+  /**
+   * When read a record batch, check whether its dictionaries are available.
+   */
+  private void checkDictionaries() throws IOException {
+    for (FieldVector vector : getVectorSchemaRoot().getFieldVectors()) {
```

Comment:
we should have a variable to indicate all dictionaries loaded so we don't need to verify this one all of them are.

Reply:
fixed.